latest/index.bs

```diff
@@ -205,25 +207,69 @@ Metadata {#metadata}
+- MUST contain the field "name" that gives the name for this dimension. The values MUST be unique across all "name" fields.
+- SHOULD contain the field "unit" to specify the physical unit of this dimension. The value SHOULD be one of the following strings, which are valid units according to UDUNITS-2.
+    - Units for "space" axes: 'angstrom', 'attometer', 'centimeter', 'decimeter', 'exameter', 'femtometer', 'foot', 'gigameter', 'hectometer', 'inch', 'kilometer', 'megameter', 'meter', 'micrometer', 'mile', 'millimeter', 'nanometer', 'parsec', 'petameter', 'picometer', 'terameter', 'yard', 'yoctometer', 'yottameter', 'zeptometer', 'zettameter'
+    - Units for "time" axes: 'attosecond', 'centisecond', 'day', 'decisecond', 'exasecond', 'femtosecond', 'gigasecond', 'hectosecond', 'hour', 'kilosecond', 'megasecond', 'microsecond', 'millisecond', 'minute', 'nanosecond', 'petasecond', 'picosecond', 'second', 'terasecond', 'yoctosecond', 'yottasecond', 'zeptosecond', 'zettasecond'
+
+If part of [[#multiscale-md]], the length of "axes" MUST be equal to the number of dimensions of the arrays that contain the image data.
+
+
+"transformations" metadata {#trafo-md}
+-------------------------------------
+
+"transformations" describes a series of transformations, e.g. to map discrete data space of an array to the corresponding physical space.
+It is a list of dictionaries. Each entry describes a single transformation and MUST contain the field "type".
+The value of "type" MUST be one of the elements of the `type` column in the table below.
+Additional fields for the entry depend on "type" and are defined by the column `fields`.
+
+| type          | fields | description |
+| ------------- | ------ |------------ |
+| `identity`    |        | identity transformation, is the default transformation and is typically not explicitly defined |
+| `translation` | one of: `"translation":List[float]`, `"path":str` | translation vector, stored either as a list of floats (`"translation"`) or as binary data at a location in this container (`path`). The length of vector defines number of dimensions. |
+| `scale`       | one of: `"scale":List[float]`, `"path":str` | scale vector, stored either as a list of floats (`scale`) or as binary data at a location in this container (`path`). The length of vector defines number of dimensions. |
+
+In addition, the field "axisIndices" MAY be given to specify the subset of axes that the transformation is applied to, leaving other axes unchanged. If not given, the transformation is applied to all axes. The length of "axisIndices" MUST be equal to the dimensionality of the transformation. If "axisIndices" are not given, the dimensionality of the transformation MUST be equal to the number of dimensions of the space that the transformation is applied to.
+If given, "axisIndices" MUST be given in increasing order. It uses zero-based indexing.
```

Comment:
This seems like a lot of rules to handle the case where some axes don't need a scale.
I wonder if instead of
`"scale": [0.5, 0.5, 0.5], "axisIndices": [2, 3, 4]` we could do `"scale": [null, null, 0.5, 0.5, 0.5]`. Then we can simply say that the length of `scale` list must be equal to the number of dimensions. And we don't need to have all the `axisIndices` rules (and validation logic). It's also less code to find the scale for e.g. dimension 2.
```
scale_dim2 = transform["scale"][2]
```
compared with:
```
index = transform["axisIndices"].indexOf(2)
if index == -1:
  scale_dim2 = None
else:
  scale_dim2 = transform["scale"][index]
```

Reply:
@will-moore - how would you know that those are the spatial axes? while a certain ordering is recommended it is not mandatory under the spec? the spec currently decouples dimension from type of axes.

Reply:
Agree that the above case is shorter, but the code to generate a 3d -> 3d transform would be longer with the 
`[null, null, 0.5, 0.5, 0.5]` case.  I.e. the code would have to check which dimensions are spatial, then slice the scale array appropriately.  

I don't have much of a preference between the two, honestly.  But some things are easier if the transforms always consider dimensions of the same type.

Reply:
You don't need `null` for scale --- you can just use 1.  And I agree that `axisIndices` just adds unnecessary complexity.

Reply:
@will-moore @jbms 

Please just keep in mind that the `axisIndices` idea was motivated by the `affine` case. i.e. to avoid having to write `[1,0,0,0,0   0,1,0,0,0   0,0,2,0,-1,  0,0,0,3,-2]` for `ctyx` for example. 

but preferring
`[2,-1,  3,-2]` with `axisIndices=[2,3]`

I hope you agree in this example, the latter is nicer.  Of course, it only matters if we need the "full affine" (i.e. if dimensions are mixed).  

I'm sympathetic to wanting to be able to get the "native resolution." of some data, in which case its good to limit the transforms to `scale` or `scale_offset`.  In which case I'm in favor of this idea.  Please consider, other more complicated use cases.  In my view, the complexity is only unnecessary when the use case is simple.

Reply:
@bogovicj It is true that it is more concise to specify an affine transform that only affects a subset of the dimensions by using `axisIndices`, but I'm not sure it is actually easier for humans to understand at a glance since then you also have to mentally perform the permutation.  But I would also expect affine transforms would be primarily read and written by computers, not by humans directly, and then it is simpler to not have to support axisIndices.

Reply:
@jbms :+1:

Mostly I'm advocating for including *some way* to describe applying transforms over subsets of data (subsets of axes).
And if it's not in this version of the spec, that's okay with me.  But if it never makes it, then we (Saalfeld lab) and others won't be able to use this spec for some of our applications, which would be a shame.

specifically - please don't make me create a 5D non-linear transform to move only the x-y axes around.

> I'm not sure it is actually easier for humans to understand at a glance then you also have to mentally perform the permutation

It helps to know that those axes are of the same type (usually spatial).  So I agree it would be awful to have to do a mental permutation, usually my "at a glance" understanding is "this affine applies to spatial dimensions only"  That more nicely matches my (John's) mental model than the 5d affine above, for example.

Actually imo, better even than `axisIndices` would be to name the axes - something you mentioned in this thread - and have these transforms apply to spaces or sub-spaces (sets of named axes).  But that's a conversation for another time and place.

I appreciate your talking through this with me.

Reply:
@bogovicj Can you explain a bit more why it is so important to be able to specify `axisIndices`?

Do you imagine frequently writing out this metadata fully manually?

With the current state of this proposal, it appears that affine transforms aren't even supported --- only translation and scale.  So there it is just a matter of specifying `0` or `1` (for translation or scale, respectively) for any axes that are to be left alone --- it seems like `axisIndices` would really only provide a benefit when manually writing out the metadata  for a full affine transform (not yet part of the proposal, but presumably planned to be added).

You mentioned the specific use case of translating x and y only.  For the use case you have in mind, is the translation by an integer amount?  If so, would that use case perhaps be better addressed by support for a non-zero origin in zarr?
https://github.com/zarr-developers/zarr-specs/issues/122

For programmatic reading and writing of this metadata, a single affine transform would seem to be the most straightforward way to deal with these transformations.  I suppose implementations that don't support anything but translation and scaling might instead normalize to a single translation vector and single scale vector.  But they could just as well use an affine transform but return an error if the rotation matrix is not diagonal.   It seems to me that composing of multiple transforms, and support for things like `axisIndices`, as specified in the current proposal, could instead by handled by library functions in the program used to generate the OME-zarr metadata.

Likewise, in a graphical user interface a single affine transform would also seem to be the most straightforward way to present the transformation, as otherwise there is a lot of added complexity in displaying possibly multiple different transforms of different types, and making it clear in what order they are applied.  In Neuroglancer affine transforms are shown in their normal matrix form but with the rows and columns labeled to indicate to which source/target dimension they correspond and which column corresponds to translation.  That makes it fairly easy to understand even if you have more than 3 dimensions.

Reply:
@bogovicj Sorry, I didn't appreciate how this would work with affine transforms. I agree that your example *with* `axisIndices` looks simpler, although I don't actually understand either!

Reply:
To quickly revive this conversation: as @bogovicj explains, we have introduced `axisIndices` to enable specifying transformations for a subset of the axes, e.g. for the use-case of affine transformations.
I think having the option of specifying subsets for transformations is crucial to have in the transformations, but I agree that `axisIndices` may not be the best solution. I see a few potential options:
1. change to using named axis for all the transformation parameters, e.g. `scale={"x": 2, "y": 2}` (as discussed above). Names must be the same as given in the axes definition. In this case we don't need `axisIndices`
2. just remove `axisIndices` from the current proposal and deal with how to address subset of axes later.
3. keep everything as is and improve how `axisIndices` are documented

My personal preference would be option 1, because I think that this would be the most elegant solution.

Reply:
Thanks for the nice summary @constantinpape .

I like option (1) as well.

> please don't make me create a 5D non-linear transform to move only the x-y axes around.

I need to clarify this, so let me describe a simpler example as a case study. There is a fairly common use case in which we need to apply a nonlinear 2D transform to every plane of a 3D volume (e.g., lens correction).

This is not covered an offset or an affine. We like the `axisIndices` (or similar) idea here because it lets us store the transform in a straightforward way: as a 2D displacement field (a 3D volume: `[2,X,Y]`, where the 2 dimension is a vector describing the displacements at `[:,x,y]`).

Without `axisIndices` (or some way to communicate "this transform is applied over a subset of the coordinates), then what is the right way to store the transform / communicate this idea? I guess by making it a 3D displacement field instead? So a `[3,X,Y,Z]` array where the `[2,:,:,:]` is full of zeros, and `[0:1,:,:,z]` all contain identical data. That's wasteful.

I'm not sure how to deal with this use case without being explicit about this "subset idea".

Reply:
I would absolutely agree that having to store extra zeros would be unacceptable, and agree that something like axisIndices is very useful in the flow field case.

You could accomplish the same thing, in a somewhat more general way, by allowing to specify an affine transform that is applied to every flow vector.  However, I'm not sure if that extra generality would actually be helpful.